### DISCUSSION TITLE
Getting a permanent redirect instead of executing the function on a URL with variables

### DISCUSSION THREAD

**Discussion by flixman** [pallets/flask#5306] [Q&A]
In my code I have the following:

```
@bp.post('/refresh', defaults={'mode': 'app'})
@bp.post('/refresh/<mode>')
def refresh_expiring_jwts(mode):
    if mode not in ['app', 'web']:
        raise NotFound
```
I would expect that when pointing a request to /refresh/app, I'd get refresh_expiring_jwts executed. Instead, I am getting back a permanent redirect to "/refresh/". I have checked the urls map and there are two rules pointing to the same method, one for /refresh and another for /refresh/<mode>. At the same time, I see that in the definition of Map (line, 99) is set to True, with this description for the parameter:
```python
    :param redirect_defaults: This will redirect to the default rule if it
                              wasn't visited that way. This helps creating
                              unique URLs.
```
This map is used in the initialization of MapAdapter, line 391, and then in then in line 628, for match():
```python
            if self.map.redirect_defaults:
                redirect_url = self.get_default_redirect(rule, method, rv, query_args)
                if redirect_url is not None:
                    raise RequestRedirect(redirect_url)
```
So... I think I am missing something here? What am I doing wrong?

Environment:

- Python version: 3.11.5
- Flask version: 3.0.0

**Comment by RonnyPfannschmidt** [comment]
If you dont have the urls in use as is,I recommend using https://werkzeug.palletsprojects.com/en/3.0.x/routing/#werkzeug.routing.AnyConverter

As for the defaults redirect, at first glance it looks like different trailing slashes might cause confusion


Personally id recommend having distinct urls and not using a default with redirect

**Reply by flixman** [reply]
I had done that, finally. What I do not understand is: what is the reason/business case for implementing this with a redirect, that can be seen from outside the application, instead of just making a call to the method?
